transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java

```diff
@@ -243,31 +250,25 @@ private static boolean isAllowHalfClosure(ChannelConfig config) {
 
     final void clearEpollIn() {
         // Only clear if registered with an EventLoop as otherwise
-        if (isRegistered()) {
-            final EventLoop loop = eventLoop();
-            final AbstractEpollUnsafe unsafe = (AbstractEpollUnsafe) unsafe();
-            if (loop.inEventLoop()) {
-                unsafe.clearEpollIn0();
-            } else {
-                // schedule a task to clear the EPOLLIN as it is not safe to modify it directly
-                loop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        if (!unsafe.readPending && !config().isAutoRead()) {
-                            // Still no read triggered so clear it now
-                            unsafe.clearEpollIn0();
-                        }
-                    }
-                });
-            }
-        } else  {
-            // The EventLoop is not registered atm so just update the flags so the correct value
-            // will be used once the channel is registered
-            flags &= ~Native.EPOLLIN;
+        final EventLoop loop = isRegistered() ? eventLoop() : null;
+        final AbstractEpollUnsafe unsafe = (AbstractEpollUnsafe) unsafe();
+        if (loop == null || loop.inEventLoop()) {
+            unsafe.clearEpollIn0();
+            return;
         }
+        // schedule a task to clear the EPOLLIN as it is not safe to modify it directly
+        loop.execute(new Runnable() {
```

Comment:
nit: couldn't we create this Runnable only once and re-use ?

Reply:
I don't think it's directly related to the change (was like this before), and it's now hit only when `setAutoRead(false)` is called from outside the EL, but sure :)

Reply:
wouldn't the old implementation make more sense since it's such an edge case? if it occurs often, sure makes sense, otherwise if it happens seldom the application will pay the price in memory for every channel.

Reply:
it's only one field so I think it is ok to cache and so reduce the GC pressure